### PR TITLE
feat: `DelayedRecordBatchStream`

### PR DESCRIPTION
**Pull request by crepererum**
Required for #5032.

### CODE REVIEW

File: datafusion_util/src/delayed.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// A [`RecordBatchStream`] that is produced by a future.
+///
+/// This can be used to delay some IO or other operations until the stream is polled.
+#[pin_project]
+pub struct DelayedRecordBatchStream<F>
```
Comment (alamb) [inline datafusion_util/src/delayed.rs:24]: `LazyRecordBatchStream` of `DeferredRecordBatchStream` would be other classic names for this pattern

**Review by alamb — APPROVED** [review]

File: datafusion_util/src/delayed.rs (hunk 1 above)
Comment (crepererum) [reply datafusion_util/src/delayed.rs:24]: will rename

**Comment by tustvold** [comment]
FWIW in DataFusion we use a combination of futures::stream::once, futures::stream::iter, try_flatten and RecordBatchStreamAdapter to avoid needing a custom abstraction

**Comment by crepererum** [comment]
@tustvold I didn't find a way to construct a stream (using existing tooling) from a future that is polled when the stream is polled and returns an error or the actual stream.

**Comment by tustvold** [comment]
@crepererum 

See https://github.com/apache/arrow-datafusion/pull/2428/files#diff-254beeb8f51dccfff404c87a3e5ebed831d62975a33e4464634c99181f2ec450R202

**Comment by crepererum** [comment]
closing because @tustvold 's suggestions works pretty well

**Pull request closed without merging**